contracts/zaps/ZapSteth.sol

```diff
@@ -0,0 +1,321 @@
+    /// @return returns the minted amounts of principal and yield tokens (PT and YT)
+    function zapStEthIn(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _ptExpected
+    ) external reentrancyGuard notFrozen returns (uint256, uint256) {
+        require(_amount != 0, "0 stETH");
+        ITranche tranche = _deriveTranche(address(_position), _expiration);
+        stETH.transferFrom(msg.sender, address(this), _amount);
+
+        uint256 balanceBegin = stETH.balanceOf(address(this));
+
+        // add liquidity with no minimum mint
+        StableSwapSTETH.add_liquidity([0, balanceBegin], 0);
+
+        uint256 outAmount = want.balanceOf(address(this));
+
+        want.transfer(address(_position), outAmount);
+
+        (uint256 ptMinted, uint256 ytMinted) = tranche.prefundedDeposit(
+            msg.sender
+        );
+
+        require(ytMinted >= outAmount, "Not enough YT minted");
+        require(ptMinted >= _ptExpected, "Not enough PT minted");
+        return (ptMinted, ytMinted);
+    }
+
+    /// @notice This function takes principal or yield tokens and redeems them for ETH.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _amountPt Amount of principal tokens to redeem.
+    /// @param _amountYt Amount of yield tokens to redeem.
+    /// @param _outputExpected Minimum amount of ETH or STETH to return
+    function zapOutEth(
+        uint256 _expiration,
+        address _position,
+        uint256 _amountPt,
+        uint256 _amountYt,
+        uint256 _outputExpected
+    ) external reentrancyGuard notFrozen {
+        _zapOut(
+            _expiration,
+            _position,
+            _amountPt,
+            _amountYt,
+            _outputExpected,
+            0
+        );
+    }
+
+    /// @notice This function takes principal or yield tokens and redeems them for stETH.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _amountPt Amount of principal tokens to redeem.
+    /// @param _amountYt Amount of yield tokens to redeem.
+    /// @param _outputExpected Minimum amount of ETH or STETH to return
+    function zapOutStEth(
+        uint256 _expiration,
+        address _position,
+        uint256 _amountPt,
+        uint256 _amountYt,
+        uint256 _outputExpected
+    ) external reentrancyGuard notFrozen {
+        _zapOut(
+            _expiration,
+            _position,
+            _amountPt,
+            _amountYt,
+            _outputExpected,
+            1
+        );
+    }
+
+    /// @notice There should never be any tokens in this contract.
+    /// This function can rescue any possible leftovers.
+    /// @param token The token to rescue.
+    /// @param amount The amount to rescue.
+    function rescueTokens(address token, uint256 amount) external onlyOwner {
```

Comment:
Let's swap it to `onlyAuthorized` and make our key management easier in the case this comes up